src/StoreBase.ts

```diff
@@ -60,178 +56,174 @@ export abstract class StoreBase {
-                        existingKeys.push(key);
-                    });
-                }
+                handleCallback(keys, sub.callback);
             });
         }
 
-        if (this._throttleMs && !StoreBase._bypassThrottle) {
-            // Needs to accumulate and trigger later -- start a timer if we don't have one running already
-            // If there are no callbacks, don't bother setting up the timer
-            if (!this._throttleTimerId && this._gatheredCallbacks.size !== 0) {
-                this._throttleTimerId = Options.setTimeout(this._resolveThrottledCallbacks, this._throttleMs);
-            }
-        } else {
-            // No throttle timeout, so just resolve now
-            this._resolveThrottledCallbacks();
+        if (!throttledUntil || bypassBlock) {
+            StoreBase._resolveCallbacks();
         }
     }
 
-    private _resolveThrottledCallbacks = () => {
-        // Prevent a store from trigginer while it's already in a trigger state
-        if (this._isTriggering) {
-            this._triggerPending = true;
+    private static _updateExistingMeta(meta: CallbackMetadata | undefined, throttledUntil: number|undefined, bypassBlock: boolean) {
+        if (!meta) {
             return;
         }
-
-        // Clear a timer if one's still pending
-        if (this._throttleTimerId) {
-            Options.clearTimeout(this._throttleTimerId);
-            this._throttleTimerId = undefined;
-            _.remove(StoreBase._pendingThrottledStores, this);
+        // Update throttling value to me min of exiting and new value
+        if (throttledUntil && meta.throttledUntil) {
+            meta.throttledUntil = Math.min(meta.throttledUntil, throttledUntil);
         }
 
-        if (StoreBase._triggerBlockCount > 0 && !this._bypassTriggerBlocks) {
-            // Trigger-blocked without a bypass flag.  Please wait until later.
-            if (!this._triggerBlocked) {
-                // Save this store to the global list that will be resolved when the block count is popped back to zero.
-                StoreBase._triggerBlockedStoreList.push(this);
-                this._triggerBlocked = true;
-            }
-            return;
+        if (!throttledUntil) {
+            meta.throttledUntil = undefined;
         }
 
-        this._triggerBlocked = false;
-        this._isTriggering = true;
-        this._triggerPending = false;
+        if (bypassBlock) {
+            meta.bypassBlock = true;
+        }
+    }
 
-        // Store the callbacks early, since calling callbacks may actually cause cascade changes to the subscription system and/or
-        // pending callbacks.
-        const storedCallbacks = this._gatheredCallbacks;
-        this._gatheredCallbacks = new Map<SubscriptionCallbackFunction, string[]>();
+    private _handleThrottledCallbacks = () => {
+        this._throttleData = undefined;
+        StoreBase._resolveCallbacks();
+    }
+    
+    private static _resolveCallbacks() {
```

Comment:
I think switching to this static handler is interesting for simplicity, but I worry that, in the case of a giant app like Skype with so much triggering flying around, that it's going to swamp this function, since it has to iterate over the entire list of callbacks for everything in the app, every time this function is called.  Once you add in the default throttle time causing EVERY trigger to be added into this map, you're looking at spending a LOT of time iterating.  I'd consider keeping basically a sorted list/linked list of callbacks sorted by triggering time, so you know when you can stop iterating.  In general I'm just worried about perf from this implementation.

Reply:
I don't think this is going to have the impact you're worried about.  Based on some testing I've done in a variety of places, over-iterating is less expensive that over-triggering here.  In practice, this shows 0-5% perf improvement in a variety of different cases.